_studio/mfx_lib/shared/src/libmfxsw_decode.cpp

```diff
@@ -495,77 +495,91 @@ mfxStatus MFXVideoDECODE_DecodeFrameAsync(mfxSession session, mfxBitstream *bs,
-            // fill dependencies
-            task.pSrc[0] = *surface_out;
-            task.pDst[0] = *surface_out;
-            // this is wa to remove external task dependency for HEVC SW decode plugin.
-            // need only because SW HEVC decode is pseudo
+            // source data is OK, go forward
+            if (task.entryPoint.pRoutine)
             {
-                mfxPlugin plugin;
-                mfxPluginParam par;
-                if (session->m_plgDec.get())
+                mfxStatus mfxAddRes;
+
+                task.pOwner = session->m_pDECODE.get();
+                task.priority = session->m_priority;
+                task.threadingPolicy = session->m_pDECODE->GetThreadingPolicy();
+                // fill dependencies
+                task.pSrc[0] = *surface_out;
+                task.pDst[0] = *surface_out;
+                // this is wa to remove external task dependency for HEVC SW decode plugin.
+                // need only because SW HEVC decode is pseudo
                 {
-                    session->m_plgDec.get()->GetPlugin(plugin);
-                    MFX_CHECK_STS(plugin.GetPluginParam(plugin.pthis, &par));
-                    if (MFX_PLUGINID_HEVCD_SW == par.PluginUID)
+                    mfxPlugin plugin;
+                    mfxPluginParam par;
+                    if (session->m_plgDec.get())
                     {
-                        task.pDst[0] = 0;
+                        session->m_plgDec.get()->GetPlugin(plugin);
+                        MFX_CHECK_STS(plugin.GetPluginParam(plugin.pthis, &par));
+                        if (MFX_PLUGINID_HEVCD_SW == par.PluginUID)
+                        {
+                            task.pDst[0] = 0;
+                        }
                     }
                 }
+
+    #ifdef MFX_TRACE_ENABLE
+                task.nParentId = MFX_AUTO_TRACE_GETID();
+                task.nTaskId = MFX::CreateUniqId() + MFX_TRACE_ID_DECODE;
+    #endif
+
+                // register input and call the task
+                mfxAddRes = session->m_pScheduler->AddTask(task, &syncPoint);
+                MFX_CHECK_STS(mfxAddRes);
             }
 
-#ifdef MFX_TRACE_ENABLE
-            task.nParentId = MFX_AUTO_TRACE_GETID();
-            task.nTaskId = MFX::CreateUniqId() + MFX_TRACE_ID_DECODE;
-#endif
+            if (MFX_ERR_MORE_DATA_SUBMIT_TASK == static_cast<int>(mfxRes))
+            {
+                mfxRes = MFX_WRN_DEVICE_BUSY;
+            }
 
-            // register input and call the task
-            mfxAddRes = session->m_pScheduler->AddTask(task, &syncPoint);
-            MFX_CHECK_STS(mfxAddRes);
-        }
+            if (MFX_WRN_DEVICE_BUSY == mfxRes)
+            {
+                mfxStatus mfxFindRes = session->m_pScheduler->FindActiveSyncPoint(session->m_pDECODE.get(), &syncPoint);
+                // If we can't find any active SyncPoint's it means that next call to MFXVideoDECODE_DecodeFrameAsync should already accept a task.
```

Comment:
When AVC decoder returns `MFX_ERR_MORE_DATA_SUBMIT_TASK` isn't the case of decoder has no active tasks and next call of 'DecodeFrameAsync' will just return the same status?

Reply:
Aha, I've recalled that we create 'virtual' (w/o output) task in that case, so the next call of `FindActiveSyncPoint` will find that sync. point. Anyway, I'd rather use sync. point that is returned by `AddTask` instead of invoking `DecodeFrameCheck` + `FindActiveSyncPoint` over again.

Reply:
IMHO this is wrong. What if AddTask wasn't called?